### PR TITLE
Validate that localhost resolves to loopback address

### PR DESCRIPTION
**Pull request by dave-fn**
## Motivation and Context
HTTP(S) credential provider requires the implementation to verify that the resolved addresses for the host are actually loopback addresses.
InetAddress is used to resolve DNS names.

## Testing
Added unit tests.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Review by millems — COMMENTED** [review]
Is there a similar change we need to make for IMDS?

File: .changes/next-release/bugfix-AmazonElasticContainerServiceECS-09f6115.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "category": "Amazon Elastic Container Service (ECS)", 
+    "contributor": "", 
+    "type": "bugfix", 
+    "description": "HTTP(S) credential provider requires the implementation to verify that the resolved addresses for the host are actually loopback addresses."
+}
```
[the remark this answers is not in the document]
Comment (dave-fn) [reply .changes/next-release/bugfix-AmazonElasticContainerServiceECS-09f6115.json:6]: Will remove.

File: core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ContainerCredentialsProvider.java [hunk 2]
```diff
@@ -207,18 +210,44 @@ private URI createUri(String relativeUri) {
...
+                                        .build();
+            }
+        }
+
+        private boolean matchesAllowedHostRules(InetAddress inetAddress) {
+            if (inetAddress instanceof Inet6Address) {
+                return ALLOWED_HOSTS_IPv6_RULES.test(inetAddress);
+            }
+
+            return ALLOWED_HOSTS_IPv4_RULES.test(inetAddress);
```
[the remark this answers is not in the document]
Comment (dave-fn) [reply core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ContainerCredentialsProvider.java:249]: Merged into a single constant. If later on, there is a need to differentiate between IPv4 and IPv6 we can create separate constants.

File: core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ContainerCredentialsProvider.java [hunk 3]
```diff
@@ -207,18 +210,44 @@ private URI createUri(String relativeUri) {
...
             return uri;
         }
+
+        private boolean isHttps(URI endpoint) {
+            return Objects.equals(HTTPS, endpoint.getScheme());
+        }
+
+        private boolean isAllowedHost(String host) {
+            try {
+                InetAddress[] addresses = InetAddress.getAllByName(host);
```
[the remark this answers is not in the document]
Comment (dave-fn) [reply core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ContainerCredentialsProvider.java:231]: Will add Javadoc.

**Comment by dave-fn** [comment]
> Is there a similar change we need to make for IMDS?

It appears IMDS can configure the endpoint, and even when not, the default values are `169.254.169.254` and `[fd00:ec2::254]`

File: core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ContainerCredentialsProvider.java [hunk 4]
```diff
@@ -207,18 +209,49 @@ private URI createUri(String relativeUri) {
...
                 String envVarName = SdkSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.environmentVariable();
                 throw SdkClientException.builder()
-                                        .message(String.format("The full URI (%s) contained within environment " +
-                                                               "variable %s has an invalid host. Host can only be one of [%s].",
-                                                               uri,
-                                                               envVarName,
-                                                               String.join(",", ALLOWED_HOSTS)))
+                                        .message(String.format("The full URI (%s) contained within environment variable " +
+                                                               "%s has an invalid host. Host should resolve to a loopback" +
+                                                               " address.",
```
Comment (millems) [inline core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ContainerCredentialsProvider.java:217]: Should we also include that making it HTTPS is an alternative way to fix it?

**Review by millems — APPROVED** [review]
LGTM, added a small non-functional comment/question. Non-blocking.

File: core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ContainerCredentialsProvider.java (hunk 4 above)
Comment (dave-fn) [reply core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ContainerCredentialsProvider.java]: Agreed. Updated exception message.

**Pull request merged**